### PR TITLE
Add support for multi-GPU inference and efficient weight loading for HuggingFace flavor

### LINKED ISSUE #8449: [FR] Add support for using HF Accelerate and efficiently loading weights
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

HuggingFace models are often large Transformer models which benefit from GPU acceleration. 

This [article](https://huggingface.co/docs/accelerate/usage_guides/big_modeling) explains why default behavior used in loading PyTorch models results in multiple copies of the model coming to exist in host RAM before the model is loaded to the GPU - this results in OOM errors even on instances where the model should be able to load.

This is because:
1. MLflow does not support sharded checkpoints during save time for models that are logged.
2. MLflow does not use Accelerate to load the model and fails to use the features that reduce memory usage
3. MLflow does not support using the device_map flag.

### Motivation

> #### What is the use case for this feature?
HuggingFace models are often large Transformer models which benefit from GPU acceleration.  Any user of LLMs with HF will benefit from this feature.

> #### Why is this use case valuable to support for MLflow users in general?

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by ankit-db**
## Related Issues/PRs

Resolve #8449 

## What changes are proposed in this pull request?

1. Support a flag to use low CPU memory by loading sharded checkpoints - We will attempt to just do this if possible. This PR adds the environment variable `MLFLOW_HUGGINGFACE_USE_LOW_CPU_MEM_USAGE` to override our default of trying to set this to true.
3. Support using device_map,  which will send shards to the GPU memory immediately and also leverage multiple GPUs if the model does not fit in a single GPU's memory. We will default to using the `auto` strategy, but we've introduced `MLFLOW_HUGGINGFACE_DEVICE_MAP_STRATEGY` to allow users to override our default of using `auto`. Other strategies are described [here](https://huggingface.co/docs/accelerate/usage_guides/big_modeling#designing-a-device-map). Note that `device` and `device_map` are not recommended to be used to together, so this setting will *override* the logic we use to automatically infer which GPU to put the model on and will also override any explicitly provided `device` passed to `load_model` via parameter or the environment variable `MLFLOW_DEFAULT_PREDICTION_DEVICE`.
4. [Deferred to future PR] Support logging with sharded checkpoints - this is a best practice, so we’re adding support for it in the spot where MLflow is calling `save_pretrained `on the pipeline that is being logged. An environment variable `MLFLOW_HUGGINGFACE_DEFAULT_WEIGHT_SHARD_SIZE` has been added to allow those who don't like this behavior to override our default of 500MB. HF's default setting is 10GB.

In order to ensure that we do not regress users who don't want accelerate in any capacity, the PR also introduces `MLFLOW_DISABLE_HUGGINGFACE_ACCELERATE_FEATURES`, which can be set to True, if a user wants to disable using any of the accelerate features for loading.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

We are verifying this works on GPU instances we launched 

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

MLflow now supports automatically using multiple GPUs when required to load large HuggingFace Transformer models and also supports loading models with strategies that reduce host memory requirements.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/transformers.py [hunk 1]
```diff
@@ -504,22 +514,22 @@ def save_model(
...
-    registered_model_name: str = None,
-    signature: Optional[ModelSignature] = None,
-    input_example: Optional[ModelInputExample] = None,
-    await_registration_for=DEFAULT_AWAIT_MAX_SLEEP_SECONDS,
-    pip_requirements: Optional[Union[List[str], str]] = None,
-    extra_pip_requirements: Optional[Union[List[str], str]] = None,
-    conda_env=None,
-    metadata: Dict[str, Any] = None,
-    **kwargs,
+        transformers_model,
```
Comment (harupy) [inline mlflow/transformers.py:517]: Can we revert unrelated changes?

File: mlflow/transformers.py [hunk 2]
```diff
@@ -836,19 +855,37 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
                 device = _TRANSFORMERS_DEFAULT_CPU_DEVICE_ID
         elif is_gpu_available():
             device = _TRANSFORMERS_DEFAULT_GPU_DEVICE_ID
+    # Note that we don't set the device in the conf yet because device is incompatible with device_map.
+    model_conf = {}
+    if MLFLOW_HUGGINGFACE_USE_DEVICE_MAP.get():
+        device_map_strategy = MLFLOW_HUGGINGFACE_DEVICE_MAP_STRATEGY.get()
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/transformers.py:861]: Ah, we use different environt varibles hhere

File: mlflow/transformers.py [hunk 3]
```diff
@@ -504,22 +514,22 @@ def save_model(
...
+        code_paths: Optional[List[str]] = None,
+        registered_model_name: str = None,
+        signature: Optional[ModelSignature] = None,
+        input_example: Optional[ModelInputExample] = None,
+        await_registration_for=DEFAULT_AWAIT_MAX_SLEEP_SECONDS,
+        pip_requirements: Optional[Union[List[str], str]] = None,
+        extra_pip_requirements: Optional[Union[List[str], str]] = None,
+        conda_env=None,
+        metadata: Dict[str, Any] = None,
+        **kwargs,
```
Comment (WeichenXu123) [inline mlflow/transformers.py:532]: Why not add argument to `save_model` method for these huggingface options ?

File: mlflow/environment_variables.py [hunk 4]
```diff
@@ -204,6 +204,36 @@ def get(self):
...
+)
+
+#: Specifies to Huggingface to use the low_cpu_mem_usage flag powered by HuggingFace accelerate.
+#: If it's set to false, the low_cpu_mem_usage flag will be set to False.
+MLFLOW_HUGGINGFACE_USE_LOW_CPU_MEM_USAGE= _EnvironmentVariable(
+    "MLFLOW_HUGGINGFACE_USE_LOW_CPU_MEM_USAGE", bool, True
+)
+
+MLFLOW_HUGGINGFACE_DEFAULT_WEIGHT_SHARD_SIZE = _EnvironmentVariable(
+    "MLFLOW_HUGGINGFACE_DEFAULT_WEIGHT_SHARD_SIZE", str, "500MB"
```
[the remark this answers is not in the document]
Comment (ankit-db) [reply mlflow/environment_variables.py:234]: It's unclear, but 10GB is a very high amount for some instances one may want to use to process weight files

There's some tradeoff with the number of produced files, but it is tunable.

File: mlflow/transformers.py (hunk 3 above)
Comment (ankit-db) [reply mlflow/transformers.py:532]: We can add it for the chunk size, but I think there should still be a default ideally

File: mlflow/environment_variables.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ankit-db) [reply mlflow/environment_variables.py]: Deferred this to another PR!

File: mlflow/environment_variables.py [hunk 5]
```diff
@@ -204,6 +204,32 @@ def get(self):
...
+# HuggingFace accelerate. If it's set to false, the low_cpu_mem_usage flag will not be
+# set to True and device_map will not be set to "auto".
+MLFLOW_DISABLE_HUGGINGFACE_ACCELERATE_FEATURES = _EnvironmentVariable(
+    "MLFLOW_DISABLE_HUGGINGFACE_ACCELERATE_FEATURES", bool, False
+)
+
+#: Specifies to Huggingface whether to use the automatic device placement logic of
+# HuggingFace accelerate. If it's set to false, the low_cpu_mem_usage flag will not be
+# set to True and device_map will not be set to "auto".
+MLFLOW_HUGGINGFACE_USE_DEVICE_MAP = _EnvironmentVariable(
```
Comment (harupy) [inline mlflow/environment_variables.py:217]:
```suggestion
MLFLOW_HUGGINGFACE_USE_DEVICE_MAP = _BooleanEnvironmentVariable(
```

We can use `_BooleanEnvironmentVariable` for a flag.

File: mlflow/transformers.py [hunk 6]
```diff
@@ -1792,7 +1823,7 @@ def trim_input(data_in, data_out):
             ):
                 # If the user has indicated to not preserve the prompt input in the response,
                 # split the response output and trim the input prompt from the response.
-                data_out = data_out[len(data_in) :].lstrip()
+                data_out = data_out[len(data_in):].lstrip()
```
Comment (harupy) [inline mlflow/transformers.py:1826]:
```suggestion
                data_out = data_out[len(data_in) :].lstrip()
```

Can we revert this unrelated change?

File: mlflow/transformers.py [hunk 7]
```diff
@@ -836,19 +852,37 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
...
+        conf[_TORCH_DTYPE_KEY] = dtype_val
+        accelerate_model_conf[_TORCH_DTYPE_KEY] = dtype_val
+
+    accelerate_model_conf["low_cpu_mem_usage"] = MLFLOW_HUGGINGFACE_USE_LOW_CPU_MEM_USAGE.get()
+
+    if not MLFLOW_DISABLE_HUGGINGFACE_ACCELERATE_FEATURES.get():
+        try:
+            model = model_instance.from_pretrained(pipeline_path,
+                                                   **accelerate_model_conf)
+        except ValueError:
```
Comment (harupy) [inline mlflow/transformers.py:880]: When does `model_instance.from_pretrained` throw `ValueError`?

File: mlflow/transformers.py [hunk 8]
```diff
@@ -96,10 +102,11 @@ def _model_packages(model) -> List[str]:
     :return: A list of strings representing the underlying engine-specific dependencies
     """
     engine = _get_engine_type(model)
+    default_deps = ["accelerate"]
```
Comment (harupy) [inline mlflow/transformers.py:105]: Do we always need `accelerate`? Tensorflow models don't need it, right?

File: mlflow/environment_variables.py [hunk 9]
```diff
@@ -204,6 +204,32 @@ def get(self):
     "MLFLOW_DEFAULT_PREDICTION_DEVICE", str, None
 )
 
+#: Specifies to Huggingface whether to use the automatic device placement logic of
+# HuggingFace accelerate. If it's set to false, the low_cpu_mem_usage flag will not be
+# set to True and device_map will not be set to "auto".
+MLFLOW_DISABLE_HUGGINGFACE_ACCELERATE_FEATURES = _EnvironmentVariable(
```
Comment (harupy) [inline mlflow/environment_variables.py:210]:
```suggestion
MLFLOW_HUGGINGFACE_DISABLE_ACCELERATE_FEATURES = _EnvironmentVariable(
```

Should this be renamed to `MLFLOW_HUGGINGFACE_DISABLE_ACCELERATE_FEATURES` because other environment variables are in the form of `MLFLOW_HUGGINGFACE_XXX_YYY`?

File: mlflow/transformers.py (hunk 7 above)
Comment (ankit-db) [reply mlflow/transformers.py:880]: Some model types don't support accelerate flags, and they throw ValueError when you pass parameters that don't exist

File: mlflow/transformers.py (hunk 6 above)
Comment (ankit-db) [reply mlflow/transformers.py:1826]: Sure - this one seems kind of minor though?

File: mlflow/transformers.py (hunk 8 above)
Comment (ankit-db) [reply mlflow/transformers.py:105]: Yeah I can move it inside the torch flag

Comment (harupy) [reply mlflow/transformers.py:105]: Sounds good :)

File: mlflow/transformers.py (hunk 7 above)
Comment (harupy) [reply mlflow/transformers.py:880]: Can you add that as a comment?

File: mlflow/transformers.py (hunk 6 above)
Comment (harupy) [reply mlflow/transformers.py:1826]: Does black format this?

Comment (harupy) [reply mlflow/transformers.py:1826]: Confirmed it does. Running black should revert this change.

**Review by harupy — APPROVED** [review]
Looks good to me once the remaining comments are addressed :) I think we should also get a reivew from Ben :)

File: mlflow/transformers.py [hunk 10]
```diff
@@ -836,19 +852,37 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
                 device = _TRANSFORMERS_DEFAULT_CPU_DEVICE_ID
         elif is_gpu_available():
             device = _TRANSFORMERS_DEFAULT_GPU_DEVICE_ID
+    # Note that we don't set the device in the conf yet because device is
+    # incompatible with device_map.
+    accelerate_model_conf = {}
+    if MLFLOW_HUGGINGFACE_USE_DEVICE_MAP.get():
+        device_map_strategy = MLFLOW_HUGGINGFACE_DEVICE_MAP_STRATEGY.get()
+        conf["device_map"] = device_map_strategy
```
Comment (BenWilson2) [inline mlflow/transformers.py:860]: Is this compatible with all Pipeline types?

File: mlflow/transformers.py (hunk 7 above)
Comment (BenWilson2) [reply mlflow/transformers.py:880]: Do TF PIpelines throw this or do they throw a Keras TypeError? We might want to wrap all of this accelerate logic around a validation of the underlying model type that was saved to ensure that we're not trying to include accelerate options on a Pipeline load for architectures that don't support it.

File: mlflow/environment_variables.py (hunk 9 above)
Comment (ankit-db) [reply mlflow/environment_variables.py:210]: Agreed - changing

File: mlflow/environment_variables.py (hunk 5 above)
Comment (ankit-db) [reply mlflow/environment_variables.py:217]: Yup - changing!

File: mlflow/transformers.py (hunk 10 above)
Comment (ankit-db) [reply mlflow/transformers.py:866]: It's not

File: mlflow/transformers.py (hunk 7 above)
Comment (ankit-db) [reply mlflow/transformers.py]: Ah I see - I guess we could default this logic to false if the engine of the underlying model is TF?

Comment (BenWilson2) [reply mlflow/transformers.py]: Yep I think that's fine. We'd have to load the `flavor_config` prior to executing `_load_model()` in order to get this logic to work and I feel as though that would be best served as a follow-up PR.

File: mlflow/environment_variables.py [hunk 11]
```diff
@@ -204,6 +204,32 @@ def get(self):
     "MLFLOW_DEFAULT_PREDICTION_DEVICE", str, None
 )
 
+#: Specifies to Huggingface whether to use the automatic device placement logic of
+# HuggingFace accelerate. If it's set to false, the low_cpu_mem_usage flag will not be
+# set to True and device_map will not be set to "auto".
+MLFLOW_HUGGINGFACE_DISABLE_ACCELERATE_FEATURES = _EnvironmentVariable(
```
Comment (BenWilson2) [inline mlflow/environment_variables.py:210]: Did we want this to be a `_BooleanEnvironmentVariable` as well?

Comment (ankit-db) [reply mlflow/environment_variables.py:210]: Great point - will do in the next update (want to see the result of tests first)

File: mlflow/transformers.py [hunk 12]
```diff
@@ -824,13 +830,36 @@ def is_gpu_available():
...
+    load_model_conf = {}
+    # Assume if torch_dtype was specified in the conf, then it must be with a pipeline for which it's compatible.
+    if _TORCH_DTYPE_KEY in conf:
+        load_model_conf[_TORCH_DTYPE_KEY] = conf[_TORCH_DTYPE_KEY]
+
+    try:
+        load_model_conf["device"] = device
+        model = model_instance.from_pretrained(pipeline_path, **load_model_conf)
+    except (ValueError, TypeError):
+        print('Could not specify device parameter for this pipeline type', flush=True)
```
Comment (BenWilson2) [inline mlflow/transformers.py:843]: Can we change this print statement to a logger warning statement?

File: mlflow/transformers.py [hunk 13]
```diff
@@ -824,13 +830,36 @@ def is_gpu_available():
         is_gpu = False
     return is_gpu
 
+def _try_load_model_with_device(model_instance, pipeline_path, device, conf):
```
Comment (BenWilson2) [inline mlflow/transformers.py:833]: We might want to rename from `pipeline_path` to `model_path`

File: mlflow/transformers.py [hunk 14]
```diff
@@ -839,19 +868,40 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
...
-    conf = {
-        "task": flavor_config[_TASK_KEY],
-        "model": model_instance.from_pretrained(pipeline_path),
-    }
     if device is not None:
         conf["device"] = device
+        accelerate_model_conf["device"] = device
+
+    if _TORCH_DTYPE_KEY in flavor_config or _TORCH_DTYPE_KEY in kwargs:
+        if _TORCH_DTYPE_KEY in kwargs:
```
Comment (BenWilson2) [inline mlflow/transformers.py:886]: +1

**Review by BenWilson2 — APPROVED** [review]
Looks solid! Great work @ankit-db :) Just some small nit comments; good to merge after they're addressed.

File: mlflow/transformers.py (hunk 12 above)
Comment (ankit-db) [reply mlflow/transformers.py:843]: Done!

**Comment by ankit-db** [comment]
@mlflow-automation autoformat
